### PR TITLE
Add dynamic matrix to refresh lockfile job

### PR DESCRIPTION
**Pull request by lbdreyer**
### Description
Creates a dynamic matrix of python version by inspecting requirements/ci/ so we don't have remember to update the matrix list.
Based off: https://docs.github.com/en/actions/learn-github-actions/expressions#example-returning-a-json-object

I've been working on this for iris-grib so I thought we may as well include it here too.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by bjlittle** [comment]
@lbdreyer Now that #4571 has landed, rebase to get the `benchmark-check` passing :+1:

File: .github/workflows/refresh-lockfiles.yml [hunk 1]
```diff
@@ -59,20 +59,30 @@ jobs:
...
     if: "github.repository == 'SciTools/iris'"
     runs-on: ubuntu-latest
-    needs: no_clobber
+    needs: get_python_matrix
     
     strategy:
       matrix:
-        python: ['38']
-    
+        python: ${{ fromJSON(needs.get_python_matrix.outputs.matrix) }}
```
Comment (trexfeathers) [inline .github/workflows/refresh-lockfiles.yml:84]: The `needs....` syntax checks out against [the documentation](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-defining-outputs-for-a-job), but shouldn't `fromJSON()` sit outside the GHA env variable? I.e:

```suggestion
        python: fromJSON(${{ needs.get_python_matrix.outputs.matrix }})
```

File: .github/workflows/refresh-lockfiles.yml [hunk 2]
```diff
@@ -59,20 +59,30 @@ jobs:
...
+    # the gen_lockfiles job.
+    if: "github.repository == 'SciTools/iris'"
+    runs-on: ubuntu-latest
+    needs: no_clobber
+    outputs:
+      matrix: ${{ steps.get_py.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v2
+      - id: get_py
+        run: echo "::set-output name=matrix::$(ls -1 requirements/ci/py??.yml | xargs -n1 basename | cut -c-4 | jq -cnR '[inputs | select(length>0)]')"
```
[the remark this answers is not in the document]
Comment (lbdreyer) [reply .github/workflows/refresh-lockfiles.yml]: Good point! I have added your suggestion to https://github.com/SciTools/iris/pull/4573/commits/7b946487863163095b0bc61231ddbebce7db5d5e

File: .github/workflows/refresh-lockfiles.yml (hunk 1 above)
Comment (lbdreyer) [reply .github/workflows/refresh-lockfiles.yml:47]: Unfortunately, when I try that github complains:
image

**Comment by lbdreyer** [comment]
Rebased to resolve a conflict due to #4618 also modifying the refresh-lockfiles.yml file

**Review by trexfeathers — APPROVED** [review]

**Comment by trexfeathers** [comment]
Thanks @lbdreyer! Sorry this fell by the wayside.

**Pull request merged**